### PR TITLE
feat(42639): "in" operator not narrowing the type if used with a const

### LINKED ISSUE #42639: "in" operator not narrowing the type if used with a const
# Bug Report

### 🔎 Search Terms

- `narrow in operator`
- `in operator`
- `narrow operator`
- `operator type guard`

### 🕗 Version & Regression Information

This is the behavior in every version I tried, and I reviewed the FAQ for entries about the "in" operator.

Tested on all TS versions available in the playground (at the time of writing: 3.3 - Nightly of 4.2.0)

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.2.0-dev.20210204&ssl=10&ssc=1&pln=11&pc=1#code/MYewdgzgLgBA1gUwJ4HcQCcAmBBGBeGAcgENCAoUSWRVDTAIXyICNyyokAHBGXAgbxgBtGmizYAugC4YYAK4BbZgnQwAvuy49GA4aLr1pMaOgCWYAObqyZTAmAAbYuh6VoMYDNwAfGPRumAGYwABQkhDDmHgCUMPxkMIkeMAD0KbLO6CAoCJgwUCAwAAbYRWQaZEGh+uKRYDFxCUnAqelgILBgmdm5ADTGUKYODsU+fmVqQA)

### 💻 Code

```ts
const keywordA = 'a'
const keywordB = 'b'

type A = { [keywordA]: number }
type B = { [keywordB]: string }

declare const c: A | B

if ('a' in c) {
    c // narrowed to `A`
}

if (keywordA in c) {
    c // not narrowed, still `A | B`
}
```

### 🙁 Actual behavior

When using a `const` string as the "prop" argument for the `in` operator the type of the object is not narrowed.

### 🙂 Expected behavior

Since I'm using a const, thus dealing with a string literal type (same as when using a literal value as an argument), I'd expect the type to be narrowed inside the `if` block.

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #42639

### CODE REVIEW

**Review by gabritto — COMMENTED** [review]
I had a minor question, but otherwise looks good to me, thanks for contributing.

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -917,7 +917,7 @@ namespace ts {
         }
 
         function isNarrowableInOperands(left: Expression, right: Expression) {
-            return isStringLiteralLike(left) && isNarrowingExpression(right);
+            return isNarrowingExpression(right) || (isIdentifier(left) || isStringLiteralLike(left));
```
[the remark this answers is not in the document]
Comment (a-tarasyuk) [reply src/compiler/binder.ts]: @gabritto Thanks for the feedback. You are right, it should be `&&`.

**Review by gabritto — APPROVED** [review]
thanks, everything seems good now

**Comment by andrewbranch** [comment]
@typescript-bot perf test this

(not sure if it works on merged PRs?)

**Comment by andrewbranch** [comment]
(It doesn’t)

**Pull request merged**